### PR TITLE
Connector refactor

### PR DESCRIPTION
**Pull request by pimeys**

### CODE REVIEW

**Comment by pimeys** [comment]
The whole `libs/sql-connection` could go away when the other engines are async.

File: introspection-engine/connectors/sql-introspection-connector/src/schema_describer_loading.rs [hunk 1]
```diff
@@ -4,17 +4,19 @@ use std::sync::Arc;
 
 pub fn load_describer(url_str: &str) -> SqlIntrospectionResult<Box<dyn SqlSchemaDescriberBackend>> {
     if url_str.starts_with("postgresql://") {
-        let wrapper = sql_connection::Postgresql::new(url_str.parse()?)?;
+        let wrapper = sql_connection::GenericSqlConnection::from_database_str(url_str, None)?;
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/src/schema_describer_loading.rs]: When asyncified, replace with `Quaint::new`

File: introspection-engine/connectors/sql-introspection-connector/src/schema_describer_loading.rs [hunk 2]
```diff
@@ -4,17 +4,19 @@ use std::sync::Arc;
...
+
         Ok(Box::new(sql_schema_describer::postgres::SqlSchemaDescriber::new(
             Arc::new(wrapper),
         )))
     } else if url_str.starts_with("mysql://") {
-        let wrapper = sql_connection::Mysql::new(url_str.parse()?)?;
+        let wrapper = sql_connection::GenericSqlConnection::from_database_str(url_str, None)?;
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/src/schema_describer_loading.rs]: When asyncified, replace with `Quaint::new`

File: introspection-engine/connectors/sql-introspection-connector/src/schema_describer_loading.rs [hunk 3]
```diff
@@ -4,17 +4,19 @@ use std::sync::Arc;
...
+
         Ok(Box::new(sql_schema_describer::mysql::SqlSchemaDescriber::new(
             Arc::new(wrapper),
         )))
     } else if url_str.starts_with("file:") {
-        let wrapper = sql_connection::Sqlite::new(url_str, "introspection-engine")?;
+        let wrapper = sql_connection::GenericSqlConnection::from_database_str(url_str, Some("introspection-engine"))?;
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/src/schema_describer_loading.rs]: When asyncified, replace with `Quaint::new`

File: introspection-engine/connectors/sql-introspection-connector/tests/test_harness/misc_helpers.rs [hunk 4]
```diff
@@ -147,7 +147,7 @@ pub fn database(database_url: &str) -> Box<dyn SyncSqlConnection + Send + Sync +
             let url = Url::parse(database_url).unwrap();
             let create_cmd = |name| format!("CREATE DATABASE \"{}\"", name);
 
-            let connect_cmd = |url| Postgresql::new(url);
+            let connect_cmd = |url: Url| GenericSqlConnection::from_database_str(url.as_str(), None);
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/tests/test_harness/misc_helpers.rs]: When asyncified, replace with `Quaint::new`.

File: introspection-engine/connectors/sql-introspection-connector/tests/test_harness/misc_helpers.rs [hunk 5]
```diff
@@ -157,13 +157,13 @@ pub fn database(database_url: &str) -> Box<dyn SyncSqlConnection + Send + Sync +
             let url = Url::parse(database_url).unwrap();
             let create_cmd = |name| format!("CREATE DATABASE `{}`", name);
 
-            let connect_cmd = |url| Mysql::new(url);
+            let connect_cmd = |url: Url| GenericSqlConnection::from_database_str(url.as_str(), None);
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/tests/test_harness/misc_helpers.rs]: When asyncified, replace with `Quaint::new`.

File: introspection-engine/connectors/sql-introspection-connector/tests/test_harness/misc_helpers.rs [hunk 6]
```diff
@@ -157,13 +157,13 @@ pub fn database(database_url: &str) -> Box<dyn SyncSqlConnection + Send + Sync +
...
 
             let conn = with_database(url, "mysql", "/", create_cmd, Rc::new(connect_cmd));
 
             Box::new(conn)
         }
-        "file" | "sqlite" => Box::new(Sqlite::new(database_url, "introspection-engine").unwrap()),
+        "file" | "sqlite" => Box::new(GenericSqlConnection::from_database_str(database_url, Some("introspection-engine")).unwrap()),
```
Comment (pimeys) [inline introspection-engine/connectors/sql-introspection-connector/tests/test_harness/misc_helpers.rs]: When asyncified, replace with `Quaint::new`.

File: libs/sql-connection/src/generic_sql_connection.rs [hunk 7]
```diff
@@ -1,117 +1,105 @@
 use datamodel::{Source, configuration::{MYSQL_SOURCE_NAME, POSTGRES_SOURCE_NAME, SQLITE_SOURCE_NAME}};
```
Comment (pimeys) [inline libs/sql-connection/src/generic_sql_connection.rs]: When asyncifying, remove all this.

File: libs/sql-connection/src/lib.rs [hunk 8]
```diff
@@ -2,20 +2,10 @@
```
Comment (pimeys) [inline libs/sql-connection/src/lib.rs]: When asyncifying, remove all this.

File: libs/sql-schema-describer/tests/sqlite/mod.rs [hunk 9]
```diff
@@ -16,7 +16,7 @@ pub fn get_sqlite_describer(sql: &str) -> sqlite::SqlSchemaDescriber {
         std::fs::remove_file(database_file_path.clone()).expect("remove database file");
     }
 
-    let conn = Sqlite::new(&format!("file://{}", database_file_path), SCHEMA).unwrap();
+    let conn = GenericSqlConnection::from_database_str(&format!("file://{}", database_file_path), Some(SCHEMA)).unwrap();
```
Comment (pimeys) [inline libs/sql-schema-describer/tests/sqlite/mod.rs]: When asyncifying, replace with `Quaint::new`.

File: migration-engine/core/src/cli.rs [hunk 10]
```diff
@@ -359,8 +359,8 @@ mod tests {
             });
 
             {
-                let uri = url::Url::parse(&mysql_url(None)).unwrap();
-                let conn = Mysql::new(uri).unwrap();
+                let uri = mysql_url(None);
+                let conn = GenericSqlConnection::from_database_str(&uri, None).unwrap();
```
Comment (pimeys) [inline migration-engine/core/src/cli.rs]: When asyncifying, replace with `Quaint::new`.

File: migration-engine/core/src/cli.rs [hunk 11]
```diff
@@ -388,8 +388,8 @@ mod tests {
             });
 
             {
-                let uri = url::Url::parse(&postgres_url(None)).unwrap();
-                let conn = Postgresql::new(uri).unwrap();
+                let uri = postgres_url(None);
+                let conn = GenericSqlConnection::from_database_str(&uri, None).unwrap();
```
Comment (pimeys) [inline migration-engine/core/src/cli.rs]: When asyncifying, replace with `Quaint::new`.

File: migration-engine/core/tests/test_harness/misc_helpers.rs [hunk 12]
```diff
@@ -42,10 +43,8 @@ pub(super) fn mysql_migration_connector(database_url: &str) -> SqlMigrationConne
         Ok(c) => c,
         Err(_) => {
             let url = Url::parse(database_url).unwrap();
-
             let name_cmd = |name| format!("CREATE DATABASE `{}`", name);
-
-            let connect_cmd = |url| Mysql::new(url);
+            let connect_cmd = |url: url::Url| GenericSqlConnection::from_database_str(url.as_str(), None);
```
Comment (pimeys) [inline migration-engine/core/tests/test_harness/misc_helpers.rs]: When asyncifying, replace with `Quaint::new`.

File: migration-engine/core/tests/test_harness/misc_helpers.rs [hunk 13]
```diff
@@ -58,8 +57,7 @@ pub(super) fn postgres_migration_connector(url: &str) -> SqlMigrationConnector {
         Ok(c) => c,
         Err(_) => {
             let name_cmd = |name| format!("CREATE DATABASE \"{}\"", name);
-
-            let connect_cmd = |url: url::Url| Postgresql::new(url);
+            let connect_cmd = |url: url::Url| GenericSqlConnection::from_database_str(url.as_str(), None);
```
Comment (pimeys) [inline migration-engine/core/tests/test_harness/misc_helpers.rs]: When asyncifying, replace with `Quaint::new`.

File: migration-engine/core/tests/test_harness/misc_helpers.rs [hunk 14]
```diff
@@ -224,18 +222,18 @@ pub fn database(sql_family: SqlFamily, database_url: &str) -> Arc<dyn SyncSqlCon
             let url = Url::parse(database_url).unwrap();
             let create_cmd = |name| format!("CREATE DATABASE \"{}\"", name);
 
-            let connect_cmd = |url| Postgresql::new(url);
+            let connect_cmd = |url: url::Url| GenericSqlConnection::from_database_str(url.as_str(), None);
```
Comment (pimeys) [inline migration-engine/core/tests/test_harness/misc_helpers.rs]: When asyncifying, replace with `Quaint::new`.

File: migration-engine/core/tests/test_harness/misc_helpers.rs [hunk 15]
```diff
@@ -224,18 +222,18 @@ pub fn database(sql_family: SqlFamily, database_url: &str) -> Arc<dyn SyncSqlCon
...
 
             let conn = with_database(url, "postgres", "postgres", create_cmd, Rc::new(connect_cmd));
 
             Arc::new(conn)
         }
-        SqlFamily::Sqlite => Arc::new(Sqlite::new(database_url, SCHEMA_NAME).unwrap()),
+        SqlFamily::Sqlite => Arc::new(GenericSqlConnection::from_database_str(database_url, Some(SCHEMA_NAME)).unwrap()),
```
Comment (pimeys) [inline migration-engine/core/tests/test_harness/misc_helpers.rs]: When asyncifying, replace with `Quaint::new`.

File: migration-engine/core/tests/test_harness/misc_helpers.rs [hunk 16]
```diff
@@ -224,18 +222,18 @@ pub fn database(sql_family: SqlFamily, database_url: &str) -> Arc<dyn SyncSqlCon
...
         SqlFamily::Mysql => {
             let url = Url::parse(database_url).unwrap();
             let create_cmd = |name| format!("CREATE DATABASE `{}`", name);
 
-            let connect_cmd = |url| Mysql::new(url);
+            let connect_cmd = |url: url::Url| GenericSqlConnection::from_database_str(url.as_str(), None);
```
Comment (pimeys) [inline migration-engine/core/tests/test_harness/misc_helpers.rs]: When asyncifying, replace with `Quaint::new`.

File: libs/sql-connection/src/connection_info.rs [hunk 17]
```diff
@@ -1,129 +0,0 @@
...
-    }
-
-    pub fn sql_family(&self) -> SqlFamily {
-        match self {
-            ConnectionInfo::Postgres(_) => SqlFamily::Postgres,
-            ConnectionInfo::Mysql(_) => SqlFamily::Mysql,
-            ConnectionInfo::Sqlite { .. } => SqlFamily::Sqlite,
-        }
-    }
-}
```
Comment (tomhoule) [inline libs/sql-connection/src/connection_info.rs]: :tada:

File: libs/sql-connection/src/generic_sql_connection.rs [hunk 18]
```diff
@@ -1,117 +1,105 @@
...
+    pub fn connection_info(&self) -> &ConnectionInfo {
+        self.pool.connection_info()
     }
 
-    fn as_sync_sql_connection(&self) -> &dyn SyncSqlConnection {
-        match self {
-            GenericSqlConnection::Postgresql(pg) => pg,
-            GenericSqlConnection::Mysql(mysql) => mysql,
-            GenericSqlConnection::Sqlite(sqlite) => sqlite,
-        }
```
Comment (tomhoule) [inline libs/sql-connection/src/generic_sql_connection.rs]: :+1:

**Pull request merged**